### PR TITLE
Added WPF + WinUI Screenshot Tool

### PR DESCRIPTION
**Pull request by duffh**
# Description

Added a developer tool to take screenshots of the WPF and WinUI sample viewer to assist with sample development. 

## Type of change

- [ ] Bug fix
- [ ] New sample implementation
- [x] Sample viewer enhancement
- [ ] Other enhancement

**Platforms tested on**:

- [x] WPF .NET 6
- [x] WPF Framework
- [x] WinUI
- [x] Xamarin.Forms Android
- [ ] Xamarin.Forms iOS
- [x] Xamarin.Forms UWP

## Checklist

- [x] Runs and compiles on all active platforms
- [x] Legacy platforms still compile and run (if applicable)
- [x] Branch is up to date with the latest main/v.next
- [x] All merge conflicts have been resolved
- [x] Self-review of changes
- [x] There are no warnings related to changes
- [x] No unrelated changes have been made to any other code or project files
- [ ] `sample_sync.py` runs without making changes
- [x] Codemaid and XAML styler extensions have been run on every changed file
- [x] Code is commented with correct formatting
- [x] All variable and method names are good and make sense
- [x] There is no leftover commented code
- [ ] Screenshots are correct size and display in description tab

### CODE REVIEW

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 1]
```diff
@@ -424,5 +456,45 @@ private void ResetCategories()
...
         }
+
+        private void ScreenshotButton_Click(object sender, RoutedEventArgs e)
+        {
+            GetJpgImage(SampleContainer);
+        }
+
+        private void GetJpgImage(UIElement source)
+        {
+            ContentControl sampleContainer = (ContentControl)source;
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:467]: These 3 lines are all unneeded.

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 2]
```diff
@@ -424,5 +456,45 @@ private void ResetCategories()
...
+            int actualHeight = (int)sampleContainer.ActualHeight;
+            int actualWidth = (int)sampleContainer.ActualWidth;
+
+            PresentationSource presentationSource = PresentationSource.FromVisual(source);
+
+            double scaleX, scaleY = 1;
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:473]: Remove this line and declare the variables in their initialization lines below.

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 3]
```diff
@@ -424,5 +456,45 @@ private void ResetCategories()
...
+            Graphics G = Graphics.FromImage(screenshot);
+            G.CopyFromScreen(X, Y, 0, 0, new System.Drawing.Size(Width, Height), CopyPixelOperation.SourceCopy);
+
+            // If scaling has occurred due to screen scaling we need to resize the image.
+            Bitmap resizedScreenshot = new Bitmap(screenshot, new System.Drawing.Size((int)(screenshot.Width / scaleX), (int)(screenshot.Height / scaleY)));
+
+            string sourcePath = ScreenshotManager.ScreenshotSettings.SourcePath;
+            string sampleName = _selectedSample.FormalName;
+            string categoryName = _selectedSample.Category;
+            string filePath = @$"{sourcePath}\{categoryName}\{sampleName}\{sampleName}.jpg"; // insert your filepath here to see the image output.
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:493]: I'm okay turning these 4 into a one-line call.
.NET Framework cannot use the `@$` string,  we have to use `$`.
```suggestion
            string filePath = $"{ScreenshotManager.ScreenshotSettings.SourcePath}\\{SampleManager.Current.SelectedSample.Category}\\{SampleManager.Current.SelectedSample.FormalName}\\{SampleManager.Current.SelectedSample.FormalName}.jpg";

```

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 4]
```diff
@@ -424,5 +456,45 @@ private void ResetCategories()
...
+            scaleY = presentationSource.CompositionTarget.TransformToDevice.M22;
+
+            int Height = (int)(source.DesiredSize.Height * scaleY);
+            int Width = (int)(source.DesiredSize.Width * scaleX);
+            System.Windows.Point relativePoint = source.PointToScreen(new System.Windows.Point(0d, 0d));
+            int X = (int)relativePoint.X;
+            int Y = (int)relativePoint.Y;
+            Bitmap screenshot = new Bitmap(Width, Height);
+            Graphics G = Graphics.FromImage(screenshot);
+            G.CopyFromScreen(X, Y, 0, 0, new System.Drawing.Size(Width, Height), CopyPixelOperation.SourceCopy);
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:485]:
```suggestion
            G.CopyFromScreen((int)relativePoint.X, (int)relativePoint.Y, 0, 0, new System.Drawing.Size(Width, Height), CopyPixelOperation.SourceCopy);
```
X and Y only get used once, we can remove them.

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 5]
```diff
@@ -424,5 +456,45 @@ private void ResetCategories()
             // Set the expanded categories.
             SetExpandedCategories(expandedCategoryNames);
         }
+
+        private void ScreenshotButton_Click(object sender, RoutedEventArgs e)
+        {
+            GetJpgImage(SampleContainer);
+        }
+
+        private void GetJpgImage(UIElement source)
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:465]: We should link to the relevant stack overflows in a comment here.
https://stackoverflow.com/q/24466482
https://stackoverflow.com/a/15537372

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 6]
```diff
@@ -424,5 +456,45 @@ private void ResetCategories()
...
+            PresentationSource presentationSource = PresentationSource.FromVisual(source);
+
+            double scaleX, scaleY = 1;
+
+            scaleX = presentationSource.CompositionTarget.TransformToDevice.M11;
+            scaleY = presentationSource.CompositionTarget.TransformToDevice.M22;
+
+            int Height = (int)(source.DesiredSize.Height * scaleY);
+            int Width = (int)(source.DesiredSize.Width * scaleX);
+            System.Windows.Point relativePoint = source.PointToScreen(new System.Windows.Point(0d, 0d));
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:480]: This point is not actually relative.
```suggestion
            System.Windows.Point screenPoint = source.PointToScreen(new System.Windows.Point(0d, 0d));
```

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs (hunk 5 above)
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:465]:
```suggestion
        private void SaveScreenshot(UIElement source)
```

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 7]
```diff
@@ -27,6 +28,7 @@ public partial class MainWindow
     {
         private bool _waitFlag;
         private List<TreeViewItem> _samples;
+        private SampleInfo _selectedSample;
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:31]: Unneeded.

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 8]
```diff
@@ -129,6 +131,8 @@ private async void categories_SelectedItemChanged(object sender, RoutedPropertyC
 
         private async Task SelectSample(SampleInfo selectedSample)
         {
+            _selectedSample = selectedSample;
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:134]: Unneeded.

File: src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs [hunk 9]
```diff
@@ -185,6 +189,13 @@ private async Task SelectSample(SampleInfo selectedSample)
 
             CategoriesRegion.Visibility = Visibility.Collapsed;
             SampleContainer.Visibility = Visibility.Visible;
+            SetScreenshotButttonVisibility(ScreenshotManager.ScreenshotSettings.ScreenshotEnabled);
+            SetContainerDimensions(ScreenshotManager.ScreenshotSettings.ScreenshotEnabled);
+        }
+
+        private void SetScreenshotButttonVisibility(bool isEnabled)
```
Comment (ZackAllen) [inline src/WPF/ArcGISRuntime.WPF.Viewer/MainWindow.xaml.cs:196]: Parameter can be removed, just reference the screenshot manager from within this method.

File: src/ArcGISRuntime.Samples.Shared/Managers/ScreenshotManager.cs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            _screenshotSettings = new ScreenshotSettings() { ScreenshotEnabled = false, SourcePath = string.Empty };
+
+            // Get the screenshot settings from the saved file if it exists.
+            // If the file does not exist, create it. 
+            if (File.Exists(Path.Combine(GetScreenshotSettingsFolder(), _screenshotSettingsFileName)))
+            {
+                string settings = File.ReadAllText(Path.Combine(GetScreenshotSettingsFolder(), _screenshotSettingsFileName));
+
+                _screenshotSettings = System.Text.Json.JsonSerializer.Deserialize(settings, typeof(ScreenshotSettings)) as ScreenshotSettings;
```
Comment (ZackAllen) [inline src/ArcGISRuntime.Samples.Shared/Managers/ScreenshotManager.cs:26]: .NET Framework does not have `System.Text.Json`. Can't compile as is.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 11]
```diff
@@ -8,17 +8,25 @@
...
 using Microsoft.UI.Xaml.Media.Imaging;
 using System;
+using System.Drawing;
 using Windows.System;
 
 namespace ArcGISRuntime.WinUI.Viewer
 {
     public sealed partial class SamplePage
     {
+        private SampleInfo _selectedSample;
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:27]: Like WPF, we can use `SampleManager.Current.SelectedSample` instead.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 12]
```diff
@@ -29,10 +37,18 @@ public SamplePage()
...
             DataContext = SampleManager.Current.SelectedSample;
+            _selectedSample = SampleManager.Current.SelectedSample;
 
             // Load and show the sample.
             SampleContainer.Content = SampleManager.Current.SampleToControl(SampleManager.Current.SelectedSample);
 
+            if (ScreenshotManager.ScreenshotSettings.ScreenshotEnabled)
+            {
+                KeyDown += ScreenshotKeyDown_Event;
+                SampleContainer.Width = ScreenshotManager.ScreenshotSettings.Width.HasValue ? ScreenshotManager.ScreenshotSettings.Width.Value : double.NaN;
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:48]: Do we know what happens if we set the value to `double.nan`?

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 13]
```diff
@@ -82,5 +98,44 @@ private async void MarkdownText_LinkClicked(object sender, LinkClickedEventArgs
...
+            int actualHeight = (int)sampleContainer.ActualHeight;
+            int actualWidth = (int)sampleContainer.ActualWidth;
+
+            double scale = ScreenshotManager.ScreenshotSettings.ScaleFactor.HasValue ? ScreenshotManager.ScreenshotSettings.ScaleFactor.Value : double.NaN;
+
+            int Height = (int)(source.DesiredSize.Height * scale);
+            int Width = (int)(source.DesiredSize.Width * scale);
+            var visual = source.TransformToVisual(null).TransformPoint(new Windows.Foundation.Point(0, 0));
+            int X = (int)(visual.X * scale);
+            int Y = (int)(visual.Y * scale) + 29;
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:122]: We should add a variable for `topBorderWidth` to keep this from being a magic number and allow for easier adjustment in the future.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 14]
```diff
@@ -8,17 +8,25 @@
...
 using System;
+using System.Drawing;
 using Windows.System;
 
 namespace ArcGISRuntime.WinUI.Viewer
 {
     public sealed partial class SamplePage
     {
+        private SampleInfo _selectedSample;
+        private Windows.System.VirtualKey _lastKeyPress;
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:28]: Unused.

File: src/ArcGISRuntime.Samples.Shared/Models/ScreenshotSettings.cs [hunk 15]
```diff
@@ -0,0 +1,11 @@
+﻿namespace ArcGISRuntime.Samples.Shared.Models
+{
+    public class ScreenshotSettings
+    {
+        public bool ScreenshotEnabled { get; set; }
+        public string SourcePath { get; set; }
```
Comment (ZackAllen) [inline src/ArcGISRuntime.Samples.Shared/Models/ScreenshotSettings.cs:6]: This path is shared between viewers, meaning we will have to change it every time we switch between viewers. We could do separate variables or handle it at the viewer level. If we store actual `src` here, we could build the path out per platform in the per platform code.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 16]
```diff
@@ -82,5 +98,44 @@ private async void MarkdownText_LinkClicked(object sender, LinkClickedEventArgs
...
+
+            double scale = ScreenshotManager.ScreenshotSettings.ScaleFactor.HasValue ? ScreenshotManager.ScreenshotSettings.ScaleFactor.Value : double.NaN;
+
+            int Height = (int)(source.DesiredSize.Height * scale);
+            int Width = (int)(source.DesiredSize.Width * scale);
+            var visual = source.TransformToVisual(null).TransformPoint(new Windows.Foundation.Point(0, 0));
+            int X = (int)(visual.X * scale);
+            int Y = (int)(visual.Y * scale) + 29;
+
+            Bitmap screenshot = new Bitmap(Width, Height);
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:124]: Most of the same feedback for WPF applies here, except we can use `@$""` in WinUI.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/ScreenshotPrompt.xaml.cs [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+            ScaleFactorEntryBox.Text = ScreenshotManager.ScreenshotSettings.ScaleFactor.HasValue ? ScreenshotManager.ScreenshotSettings.ScaleFactor.ToString() : null;
+        }
+
+        private void SaveButton_Click(object sender, RoutedEventArgs e)
+        {
+            ScreenshotSettings screenshotSettings = new ScreenshotSettings();
+            screenshotSettings.ScreenshotEnabled = ScreenshotEnabledCheckBox.IsChecked.HasValue ? ScreenshotEnabledCheckBox.IsChecked.Value : false;
+            screenshotSettings.SourcePath = SourcePathText.Text;
+
+            if (int.TryParse(WidthEntryBox.Text, out int width))
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/ScreenshotPrompt.xaml.cs:48]: I think try parse is sufficient with this being an internal tool. The regex check might be good for some samples in the future, but it is not needed for this.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 18]
```diff
@@ -82,5 +98,44 @@ private async void MarkdownText_LinkClicked(object sender, LinkClickedEventArgs
...
+            Graphics G = Graphics.FromImage(screenshot);
+            G.CopyFromScreen(X, Y, 0, 0, new System.Drawing.Size(Width, Height), CopyPixelOperation.SourceCopy);
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:126]:
```suggestion
            G.CopyFromScreen(X, Y, 0, 0, new Size(Width, Height), CopyPixelOperation.SourceCopy);
```

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs [hunk 19]
```diff
@@ -82,5 +98,44 @@ private async void MarkdownText_LinkClicked(object sender, LinkClickedEventArgs
...
+        {
+            if (e.Key == Windows.System.VirtualKey.End)
+            {
+                GetJpgImage(SampleContainer);
+            }
+        }
+
+        private void GetJpgImage(UIElement source)
+        {
+            ContentControl sampleContainer = (ContentControl)source;
```
Comment (ZackAllen) [inline src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:112]: These 3 lines are unneeded.

**Review by ZackAllen — CHANGES_REQUESTED** [review]

**Comment by ZackAllen** [comment]
WinUI only works on the primary display. I am okay with this.

File: src/ArcGISRuntime.Samples.Shared/Models/ScreenshotSettings.cs (hunk 15 above)
Comment (duffh) [reply src/ArcGISRuntime.Samples.Shared/Models/ScreenshotSettings.cs:6]: Good point and well spotted, I think storing the src here and building up per platform is the correct solution.

File: src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs (hunk 12 above)
Comment (duffh) [reply src/WinUI/ArcGISRuntime.WinUI.Viewer/SamplePage.xaml.cs:48]: It works as intended, it resets the container back to the default size.

**Review by ZackAllen — APPROVED** [review]

**Pull request merged**